discovery/pkg/metrics/metrics.go

```diff
@@ -109,65 +172,143 @@ func (d *DiscovererMetrics) RegisterPrometheus() {
 	if ok {
-		m.WithLabelValues(backendName, clusterType, path).Set(math.Floor(duration.Seconds() * 1e3))
+		m.WithLabelValues(d.backendName, d.backendType, path).Set(math.Floor(duration.Seconds() * 1e3))
+	}
+}
+
+// DiscovererUpstreamServicesMetric records the total number of upstream services
+func (d *DiscovererMetrics) DiscovererUpstreamServicesMetric(namespace string, totalServices int) {
+	m, ok := d.metrics[DiscovererUpstreamServicesGauge].(*prometheus.GaugeVec)
+	if ok {
+		m.WithLabelValues(d.backendName, namespace, d.backendType).Set(float64(totalServices))
+	}
+}
+
+// DiscovererReplicatedServicesMetric records the total number of replicated services
+func (d *DiscovererMetrics) DiscovererReplicatedServicesMetric(namespace string, totalServices int) {
+	m, ok := d.metrics[DiscovererReplicatedServicesGauge].(*prometheus.GaugeVec)
+	if ok {
+		m.WithLabelValues(d.backendName, namespace, d.backendType).Set(float64(totalServices))
+	}
+}
+
+// DiscovererInvalidServicesMetric records the total number of invalid services
+func (d *DiscovererMetrics) DiscovererInvalidServicesMetric(namespace string, totalServices int) {
+	m, ok := d.metrics[DiscovererInvalidServicesGauge].(*prometheus.GaugeVec)
+	if ok {
+		m.WithLabelValues(d.backendName, namespace, d.backendType).Set(float64(totalServices))
+	}
+}
+
+// DiscovererUpstreamEndpointsMetric records the total upstream endpoints in the backend cluster
+func (d *DiscovererMetrics) DiscovererUpstreamEndpointsMetric(namespace, serviceName string, totalEp int) {
+	m, ok := d.metrics[DiscovererUpstreamEndpointsGauge].(*prometheus.GaugeVec)
+	if ok {
+		m.WithLabelValues(d.backendName, namespace, serviceName, d.backendType).Set(float64(totalEp))
+	}
+}
+
+// DiscovererReplicatedEndpointsMetric records the total replicated endpoints
+func (d *DiscovererMetrics) DiscovererReplicatedEndpointsMetric(namespace, serviceName string, totalEp int) {
+	m, ok := d.metrics[DiscovererReplicatedEndpointsGauge].(*prometheus.GaugeVec)
+	if ok {
+		m.WithLabelValues(d.backendName, namespace, serviceName, d.backendType).Set(float64(totalEp))
+	}
+}
+
+// DiscovererInvalidEndpointsMetric records the total replicated endpoints
+func (d *DiscovererMetrics) DiscovererInvalidEndpointsMetric(namespace, serviceName string, totalEp int) {
+	m, ok := d.metrics[DiscovererInvalidEndpointsGauge].(*prometheus.GaugeVec)
+	if ok {
+		m.WithLabelValues(d.backendName, namespace, serviceName, d.backendType).Set(float64(totalEp))
+	}
+}
+
+// DiscovererInfoMetric records version information
+func (d *DiscovererMetrics) DiscovererInfoMetric(version string) {
+	m, ok := d.metrics[DiscovererInfoGauge].(*prometheus.GaugeVec)
+	if ok {
+		m.WithLabelValues(d.backendName, version, d.backendType).Set(1)
+	}
+}
+
+// GetTotalServicesCount returns the number of services in a namespace for the particular backend
+func (d *DiscovererMetrics) GetTotalServicesCount(kubeclient kubernetes.Interface, namespace string) (int, error) {
```

Comment:
These funcs seem somewhat out of place in that they are unrelated to the metrics package. Wondering if it makes sense to move these calls to the k8s API into the sync pkg, which is where this is called and it already has a dependency on the k8s packages.

Reply:
I needed a good home for this but couldn't find one. Also, I need access to `backendName` to filter on labels. If it's moved from metrics will just need to make those params public.

Reply:
Aaaaah I see. I think another option might be to add a `backendName` field to `endpointsAction` and `serviceAction`, and do `AddEndpointsAction(backendName, endpoints)`. The downside is that we have to plumb `backendName` again through a couple of layers. Not sure if it's worth it at this time.

Reply:
I moved stuff around, just picked the backend name out of metrics, PTAL